doc/developer/design/20231110_aws_connections.md

```diff
@@ -0,0 +1,164 @@
+handle the AssumeRole scenario as well.
+
+In the parser we already support `ROLE ARN` option which can be renamed to `ASSUME ROLE ARN`
+to better reflect the usage. We will need to add an `ASSUME ROLE SESSION NAME` option as well.
+
+The AWS Connection create statement with all the options will now look like:
+```sql
+CREATE CONNECTION <connector_name> TO AWS (
+  ACCESS KEY ID [[=] <value>], -- existing
+  SECRET ACCESS KEY [[=] SECRET <value>], -- existing
+  TOKEN [[=] SECRET <value>], -- existing
+  ASSUME ROLE ARN [[=] <value>], -- rename existing from ROLE ARN
+  ASSUME ROLE SESSION NAME [[=] <value>], -- new option
+  ENDPOINT [[=] <value>], -- existing
+  REGION [[=] <value>] -- existing
+);
+```
+
+Users should be able to provide either the `ACCESS KEY` options or the `ASSUME ROLE` options,
+but not both.
+
+#### External ID and principal
+We already have external ID prefix provided in the catalog state.
+The connection ID will be appended to the external ID prefix to get the complete
+External ID and then stored in the new `mz_aws_connections` table for that connection ID.
+For principal, refer to the [open question below](#what-should-be-the-principal-for-this-connection).
+
+#### `VALIDATE CONNECTION`
+We can make use of the [GetCallerIdentity AWS API](https://docs.aws.amazon.com/STS/latest/APIReference/API_GetCallerIdentity.html)
+to check if the credentials are valid. We should make the call with the given External ID and without
+to make sure that the user has a secure setup. If we are able to access without the External ID,
+we should treat that as an error.
+
+Optionally, with `ASSUME ROLE ARN` we can also try to get temporary credentials.
+
+## Rollout and Testing
+We should put the AWS connection behind a feature flag.
+
+#### Testing During development
+Write cloudtests to test out the different AWS Connections. For the AssumeRole AWS
+Connection we should test that the expected principal and external_id is populated
+in the `mz_aws_connections` table.
+
+#### Testing after code is merged
+Switch on the feature flag for the staging environment and create an AWS Connection
+in staging followed by `VALIDATE CONNECTION`.
+
+## Future work
+
+#### Extend `VALIDATE CONNECTION` of an AWS connection to be comprehensive.
+An AWS connection can be potentially re-used across multiple services like S3 or RDS.
+Having permission to an S3 bucket might not mean that RDS access is set up correctly. So
+a validate connection will not be comprehensive and we'll probably do some additional check
+with the intended resources when we actually make use of this connection.
+
+Hence, potentially later we can extend `VALIDATE CONNECTION` sql to optionally specify an S3 or a database url,
+like, `VALIDATE CONNECTION aws_conn WITH (S3 PATH = 's3://prefix')` so the user can correctly
+validate if the policies are set up correctly on their end in AWS.
```

Comment:
@moulimukherjee This makes sense to me -- a good follow on. What does the `VALIDATE CONNECTION` return, just success/failure? Can it return anything about the credentials or role used to establish the connection? I can imagine one user creating the connection for copy to S3 and another user wants to check whether they can re-use it. Could they get the information they need with `VALIDATE` ? Or the connections table?

Reply:
From the materialize docs, 
> A successful connection and authentication attempt returns successfully with no rows. If an issue is encountered, the command will return a validation error.

We can probably keep the same behaviour and users can check using `VALIDATE CONNECTION ... WITH ...` to see if it still works with a different AWS resource.

Reply:
It's possible we don't actually need this to be built in to `VALIDATE CONNECTION`. In fact, I'm not even sure it's possible. Is it possible to verify that you have `PutObject` permissions without just _trying_ to do a `PutObject` request? I'm not sure that it is. Of course you could put _and_ remove an object, as you suggested below—but there's a chance that we've been granted `PutObject` but _not_ `RemoveObject` permissions, and now `VALIDATE CONNECTION` has just leaked an object. I don't love the idea of `VALIDATE CONNECTION` leaking state like that. Not fundamentally opposed—just something to think through.

Alternative take: it might be sufficient to simply do this validation at the time someone does `CREATE SOURCE` or `COPY ... TO S3` and provide good error messages! An easy way to guarantee that you have permission to do a `COPY ... TO S3` is to ... do a `COPY ... TO S3`. We could write some docs that say "to validate your AWS connection actually has S3 permissions, try exporting a very simple relation:"

```
COPY (SELECT 1) TO S3 USING <the AWS connection you just created>;
```

Reply:
@benesch That's a good point about `PutObject` and `RemoveObject`. We can do comprehensive validation while doing the actual `COPY`.

Reply:
One note though, for `COPY ... TO S3` I think we'll need both `PutObject` and `RemoveObject` permissions. If there was data partially written to the prefix, we should allow like an overwrite option which empties the directory and starts the export again. We can revisit this with the copy to s3 design doc.

Reply:
> We should allow like an overwrite option 
Yes -- definitely want this!

Reply:
> If there was data partially written to the prefix or the directory is not empty, the copy should fail.

I'm not so sure! It's worth looking at what Snowflake does. See https://docs.snowflake.com/en/user-guide/data-load-s3-config-aws-iam-user#aws-access-control-requirements. I _think_ the only permission required for unloading (their equivalent of `COPY TO`) is `s3:PutObject`. There might be a good reason for that. Perhaps it's a hassle to try to get users to grant `s3:ListObject` in addition to `s3:PutObject`, since it's not strictly necessary.

> We should allow like an overwrite option which empties the directory before exporting the data again.

Snowflake's `OVERWRITE` option has interesting semantics:

> Boolean that specifies whether the COPY command overwrites existing files with matching names, if any, in the location where files are stored. The option does not remove any existing files that do not match the names of the files that the COPY command unloads.
> 
> In many cases, enabling this option helps prevent data duplication in the target stage when the same COPY INTO <location> statement is executed multiple times. However, when an unload operation writes multiple files to a stage, Snowflake appends a suffix that ensures each file name is unique across parallel execution threads (e.g. data_0_1_0). The number of parallel execution threads can vary between unload operations. If the files written by an unload operation do not have the same filenames as files written by a previous operation, SQL statements that include this copy option cannot replace the existing files, resulting in duplicate files.
> 
> In addition, in the rare event of a machine or network failure, the unload job is retried. In that scenario, the unload operation writes additional files to the stage without first removing any files that were previously written by the first attempt.
> 
> To avoid data duplication in the target stage, we recommend setting the INCLUDE_QUERY_ID = TRUE copy option instead of OVERWRITE = TRUE and removing all data files in the target stage and path (or using a different path for each unload operation) between each unload job.

Now, I think the semantics there are awfully confusing. Probably so confusing that we wouldn't want to support `OVERWRITE` with such semantics. But I think it's notably how careful they are to never remove files from an S3 bucket—only possibly overwrite the existing files.

I think we likely want to do the same. Otherwise there is a possibility of catastrophe that I don't think we want to be responsible for. Imagine that you mean to run

```sql
COPY ...  TO 's3://my-data-export/2023/04/03/my-team-export' (OVERWRITE = TRUE)
```

to export your team's daily data to the `my-data-export` bucket. But say you absentmindly type instead:

```sql
`COPY ... TO 's3://my-data-export' (OVERWRITE = TRUE)`
```

If `OVERWRITE` first empties the prefix ... this command will _completely_ wipe out the entire `my-data-export` bucket. Terrifying. 

I think it's much safer to say it's the user's responsibility to ensure that the target prefix of S3 is empty before the export begins. We can offer something like the `INCLUDE_QUERY_ID` option to automatically append a UUID to the provided path, to ensure uniqueness.

Reply:
Yeah, Snowflake actually hard-errors if the prefix already has existing files (I think they do require `ListBucket` for this). And fat fingering with the wrong prefix does sound terrifying (even if versioning is enabled, recovering the previous state is not pretty). I like the query ID approach to always have a unique prefix or at least give it as an option instead of just erroring out if somebody wants to re-run a failed COPY.

Reply:
I updated the section to mention using something like `COPY (SELECT 1) TO S3 USING <the AWS connection you just created>;` instead.

Reply:
> Yeah, Snowflake actually hard-errors if the prefix already has existing files (I think they do require ListBucket for this)

What if you don't grant Snowflake `ListObjects` permission? Do you think it just skips the check?

Reply:
Good question, let me check this out.

Reply:
@benesch I got an access denied without the `ListBucket` even though the `PutObject` and other access were there.